### PR TITLE
Implement bank client

### PR DESCRIPTION
**Pull request by DucPhamTV**
- Create BaseClient class.
- Implement Bank class.
- Write unit tests for Bank class.

Relative issues: #3 , #4 , #5 , #6 , #7 , #8 .

### CODE REVIEW

**Comment by shahraizali** [comment]
Okay, cool will review it today. just looking at the code looks like a great job done 👍 . I'll pull the code and test it further. @buckyroberts  if you have some time today please do give your feedback as well at least in a couple of initial PRs so we are on a good start with the structure of the code and developers usually feel better jumping in if the structure is already laid out.

**Comment by shahraizali** [comment]
Hi @DucPhamTV ! I have tested issues #3, #5 #6 #8  with hitting a live server and the response is correct, great job there 👍 , btw these are all fetch requests so didn't require any data to be sent but for issues #7 and #4 those require data to be sent along. did you test them using actual signatures and node id etc?  Also, I think we need to describe a bit more in the comments for example about `node_id`, `trust`, and `signature` like whose signature is required and whose node_id, etc Thoughts?  @DucPhamTV @buckyroberts

**Comment by DucPhamTV** [comment]
Hi @shahraizali ,

I'm looking around to test it with actual Bank server. Since #4 and #7 are PATCH and POST requests, I'm afraid it would have some side affect to the server. So I intend to start my own Bank server to do the manual tests.

I also have a question about node_identifier. Is it the node_identifier of the Bank server or the Bank client? I read this https://thenewboston.com/guide/node-identifiers I understood that it is node_identifier of bank client. I open Account Manager to look for it but haven't found it yet.

**Comment by shahraizali** [comment]
Yeah, It's a good idea not to ping the server that why I asked you. I have asked bucky he said we can just go with the mocks for now but we can definitely set up our local test server. He will be responding to the node identifier shortly!

**Comment by buckyroberts** [comment]
Every node (Bank, CV, and PV) have a `node_identifier` that they can use to prove that the owner of that server has authorized the request. However this request can come from the client or server. For example, when you are the owner of a Bank and you want to send a connection request to a CV you can either:

Send it from the Bank
`Bank --> CV`

Or send it from a client app (or even Postman)
`Client ---> CV`

In both cases, that CV is going to connect to the Bank to make sure that the `signature` on the request is valid for that `node_identifier`. So not so important where the request came from in that example, it's just important that the request has been authorized by the owner of that node.

**Comment by shahraizali** [comment]
@DucPhamTV so it's the bank's identifier here since we are in the bank class. so the bank has to give that proof that he owns the right to update stuff and make the patch call. makes sense right?

File: tnb/banks.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import requests
```
Comment (shahraizali) [inline tnb/banks.py:1]: extra import

File: tests/banks.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+import pytest
+import requests_mock
```
Comment (shahraizali) [inline tests/banks.py:2]: we can remove these extra imports other than that looks like you nailed it 👍

**Comment by DucPhamTV** [comment]
Thanks for clear explanation @buckyroberts !

Hi @shahraizali To verify the issue #4 anh #7, I tried to start a bank node on my machine. If I follow Bank's README. It seems only for development, the tests can run successfully, but I can't access to 127.0.0.1:8000 to do some requests. Does it only production environment allow clients to interact with the node? If yes, I will follow up this guide https://thenewboston.com/deployment-guide/bank. But when the bank is up, does it requires PV and CV to work? :thinking:

I can get bank's `node_identifier` from `/config`. About `signature`, To generate it, beside `node_identifier` as its public key, If I'm able to start a Bank node in production, I can add `node_identifier` at the step `initialize_bank` and of course I have signing key in hand. But I don't know the format message to be signed and the node_identifier signing key. I'm looking into the code...

I'm sorry but a lot of questions in my mind now.

**Comment by shahraizali** [comment]
Great going 👍 for local you can use this command to populate the configs it will add the dump note that your local data will be overwritten for 3 models ('validator', 'self_configuration', 'user')

command : `python manage.py initialize_test_bank`

make sure you have variables exported
export DJANGO_APPLICATION_ENVIRONMENT='local'
export NETWORK_SIGNING_KEY='e5e5fec0dcbbd8b0a76c67204823678d3f243de7a0a1042bb3ecf66285cd9fd4'

**Comment by shahraizali** [comment]
The bank does require the celery beat to be up like mentioned in the docs. we can go ahead with the mock data as well you can just give it a final look and I'll merge after that looks fine to me. bucky will be testing it further.

**Comment by DucPhamTV** [comment]
Thanks @shahraizali .I've done my final check. Please have a look on it.

**Comment by shahraizali** [comment]
@angle943 Issues to be paid: #3 , #4 , #5 , #6 , #7 , #8 .

**Comment by DucPhamTV** [comment]
My account number ec5aa47719460ee5bcd16ab3cfe94c521d702eb9fc7d80f603762c23014f8ab8

**Comment by odetolataiwo** [comment]
paid

**Pull request merged**